diff --git a/Makefile b/Makefile
index 1222d8c..3daf9b3 100644
--- a/Makefile
+++ b/Makefile
@@ -5,8 +5,13 @@
 gb  := gb
 gba := gba
 
-profile := accuracy
-target := higan
+ifndef profile
+  profile := accuracy
+endif
+
+ifndef target
+  target := higan
+endif
 # target := loki
 
 ifeq ($(target),loki)
@@ -29,7 +31,8 @@ ifeq ($(pgo),instrument)
   flags += -fprofile-generate
   link += -lgcov
 else ifeq ($(pgo),optimize)
-  flags += -fprofile-use
+  flags += -fprofile-use -fprofile-correction
+  link += -lgcov
 endif
 
 # platform

--- a/target-higan/Makefile	2015-01-25 14:50:37.405869722 +0900
+++ b/target-higan/Makefile	2015-01-25 14:59:58.857311200 +0900
@@ -1,4 +1,11 @@
-name := higan
+ifndef name
+  name := higan
+endif
+
+ifndef DESTDIR
+  DESTDIR := /
+endif
+
 
 processors := arm gsu hg51b lr35902 r6502 r65816 spc700 upd96050
 include processor/Makefile
@@ -89,29 +96,25 @@
 	sourcery $(ui)/resource/resource.bml $(ui)/resource/resource.cpp $(ui)/resource/resource.hpp
 
 install:
-ifneq ($(shell id -un),root)
-	$(error "make install must be run as root")
-else ifeq ($(platform),windows)
+ifeq ($(platform),windows)
 else ifeq ($(platform),macosx)
-	mkdir -p /Library/Application\ Support/$(name)
-	cp -R profile/* /Library/Application\ Support/$(name)
-	cp data/cheats.bml /Library/Application\ Support/$(name)/cheats.bml
-	chmod -R 777 /Library/Application\ Support/$(name)
+	mkdir -p $(DESTDIR)Library/Application\ Support/$(name)
+	cp -R profile/* $(DESTDIR)Library/Application\ Support/$(name)
+	cp data/cheats.bml $(DESTDIR)Library/Application\ Support/$(name)/cheats.bml
+	chmod -R 777 $(DESTDIR)Library/Application\ Support/$(name)
 else
-	cp out/$(name) $(prefix)/bin/$(name)
-	cp data/$(name).png $(prefix)/share/pixmaps/$(name).png
-	cp data/$(name).desktop $(prefix)/share/applications/$(name).desktop
-	mkdir -p /usr/share/$(name)
-	cp -R profile/* /usr/share/$(name)
-	cp data/cheats.bml /usr/share/$(name)/cheats.bml
-	chmod -R 777 /usr/share/$(name)
+	cp out/$(name) $(DESTDIR)$(prefix)/bin/$(name)
+	cp data/$(name).png $(DESTDIR)$(prefix)/share/pixmaps/$(name).png
+	cp data/$(name).desktop $(DESTDIR)$(prefix)/share/applications/$(name).desktop
+	mkdir -p $(DESTDIR)usr/share/$(name)
+	cp -R profile/* $(DESTDIR)usr/share/$(name)
+	cp data/cheats.bml $(DESTDIR)usr/share/$(name)/cheats.bml
+	chmod -R 777 $(DESTDIR)usr/share/$(name)
 endif
 
 uninstall:
-ifneq ($(shell id -un),root)
-	$(error "make uninstall must be run as root")
-else ifeq ($(platform),windows)
+ifeq ($(platform),windows)
 else ifeq ($(platform),macosx)
 else
-	rm $(prefix)/bin/$(name)
+	rm $(DESTDIR)$(prefix)/bin/$(name)
 endif
--- a/target-loki/Makefile	2015-01-25 13:10:13.465004978 +0900
+++ b/target-loki/Makefile	2015-01-25 15:01:39.198002214 +0900
@@ -1,4 +1,10 @@
-name := loki
+ifndef name
+  name := loki
+endif
+
+ifndef DESTDIR
+  DESTDIR := /
+endif
 
 processors := arm gsu hg51b lr35902 r65816 spc700 upd96050
 include processor/Makefile
@@ -72,25 +78,21 @@
 	sourcery $(ui)/resource/resource.bml $(ui)/resource/resource.cpp $(ui)/resource/resource.hpp
 
 install:
-ifneq ($(shell id -un),root)
-	$(error "make install must be run as root")
-else ifeq ($(platform),windows)
+ifeq ($(platform),windows)
 else ifeq ($(platform),macosx)
-	mkdir -p /Library/Application\ Support/$(name)
-	cp -R profile/* /Library/Application\ Support/$(name)
-	chmod -R 777 /Library/Application\ Support/$(name)
+	mkdir -p $(DESTDIR)Library/Application\ Support/$(name)
+	cp -R profile/* $(DESTDIR)Library/Application\ Support/$(name)
+	chmod -R 777 $(DESTDIR)Library/Application\ Support/$(name)
 else
-	cp out/$(name) $(prefix)/bin/$(name)
-	mkdir -p /usr/share/$(name)
-	cp -R profile/* /usr/share/$(name)
-	chmod -R 777 /usr/share/$(name)
+	cp out/$(name) $(DESTDIR)$(prefix)/bin/$(name)
+	mkdir -p $(DESTDIR)usr/share/$(name)
+	cp -R profile/* $(DESTDIR)usr/share/$(name)
+	chmod -R 777 $(DESTDIR)usr/share/$(name)
 endif
 
 uninstall:
-ifneq ($(shell id -un),root)
-	$(error "make uninstall must be run as root")
-else ifeq ($(platform),windows)
+ifeq ($(platform),windows)
 else ifeq ($(platform),macosx)
 else
-	rm $(prefix)/bin/$(name)
+	rm $(DESTDIR)$(prefix)/bin/$(name)
 endif
