diff --git a/Makefile b/Makefile
index 1222d8c..3daf9b3 100644
--- a/Makefile
+++ b/Makefile
@@ -5,9 +5,13 @@ sfc := sfc
 gb  := gb
 gba := gba
 
-profile := accuracy
-target := higan
-# target := loki
+ifndef profile
+  profile := accuracy
+endif
+
+ifndef target
+  target := higan
+endif
 
 ifeq ($(target),loki)
   options += debugger
@@ -29,7 +31,8 @@ ifeq ($(pgo),instrument)
   flags += -fprofile-generate
   link += -lgcov
 else ifeq ($(pgo),optimize)
-  flags += -fprofile-use
+  flags += -fprofile-use -fprofile-correction
+  link += -lgcov
 endif
 
 # platform

--- a/target-higan/Makefile	2015-01-25 14:50:37.405869722 +0900
+++ b/target-higan/Makefile	2015-01-25 14:59:58.857311200 +0900
@@ -1,4 +1,11 @@
-name := higan
+ifndef name
+  name := higan
+endif
+
+ifndef DESTDIR
+  DESTDIR := /
+endif
+
 
 processors := arm gsu hg51b lr35902 r6502 r65816 spc700 upd96050
 include processor/Makefile
@@ -89,29 +96,28 @@
 	sourcery $(ui)/resource/resource.bml $(ui)/resource/resource.cpp $(ui)/resource/resource.hpp
 
 install:
-ifneq ($(shell id -un),root)
-	$(error "make install must be run as root")
-else ifeq ($(platform),windows)
+ifeq ($(platform),windows)
 else ifeq ($(platform),macosx)
-	mkdir -p /Library/Application\ Support/$(name)
-	cp -R profile/* /Library/Application\ Support/$(name)
-	cp data/cheats.bml /Library/Application\ Support/$(name)/cheats.bml
-	chmod -R 777 /Library/Application\ Support/$(name)
+	install -d $(DESTDIR)Library/Application\ Support/$(name)
+	cp -R profile/* $(DESTDIR)Library/Application\ Support/$(name)
+	cp data/cheats.bml $(DESTDIR)Library/Application\ Support/$(name)/cheats.bml
+	#chmod -R 777 $(DESTDIR)Library/Application\ Support/$(name)
 else
-	cp out/$(name) $(prefix)/bin/$(name)
-	cp data/$(name).png $(prefix)/share/pixmaps/$(name).png
-	cp data/$(name).desktop $(prefix)/share/applications/$(name).desktop
-	mkdir -p /usr/share/$(name)
-	cp -R profile/* /usr/share/$(name)
-	cp data/cheats.bml /usr/share/$(name)/cheats.bml
-	chmod -R 777 /usr/share/$(name)
+	install -d $(DESTDIR)$(prefix)/bin/
+	cp out/$(name) $(DESTDIR)$(prefix)/bin/$(name)
+	install -d $(DESTDIR)$(prefix)/share/pixmaps/
+	cp data/$(name).png $(DESTDIR)$(prefix)/share/pixmaps/$(name).png
+	install -d $(DESTDIR)$(prefix)/share/applications/
+	cp data/$(name).desktop $(DESTDIR)$(prefix)/share/applications/$(name).desktop
+	install -d $(DESTDIR)usr/share/$(name)
+	cp -R profile/* $(DESTDIR)usr/share/$(name)
+	cp data/cheats.bml $(DESTDIR)usr/share/$(name)/cheats.bml
+	#chmod -R 777 $(DESTDIR)usr/share/$(name)
 endif
 
 uninstall:
-ifneq ($(shell id -un),root)
-	$(error "make uninstall must be run as root")
-else ifeq ($(platform),windows)
+ifeq ($(platform),windows)
 else ifeq ($(platform),macosx)
 else
-	rm $(prefix)/bin/$(name)
+	rm $(DESTDIR)$(prefix)/bin/$(name)
 endif
--- a/target-loki/Makefile	2015-01-25 13:10:13.465004978 +0900
+++ b/target-loki/Makefile	2015-01-25 15:01:39.198002214 +0900
@@ -1,4 +1,10 @@
-name := loki
+ifndef name
+  name := loki
+endif
+
+ifndef DESTDIR
+  DESTDIR := /
+endif
 
 processors := arm gsu hg51b lr35902 r65816 spc700 upd96050
 include processor/Makefile
@@ -72,25 +78,22 @@
 	sourcery $(ui)/resource/resource.bml $(ui)/resource/resource.cpp $(ui)/resource/resource.hpp
 
 install:
-ifneq ($(shell id -un),root)
-	$(error "make install must be run as root")
-else ifeq ($(platform),windows)
+ifeq ($(platform),windows)
 else ifeq ($(platform),macosx)
-	mkdir -p /Library/Application\ Support/$(name)
-	cp -R profile/* /Library/Application\ Support/$(name)
-	chmod -R 777 /Library/Application\ Support/$(name)
+	install -d $(DESTDIR)Library/Application\ Support/$(name)
+	cp -R profile/* $(DESTDIR)Library/Application\ Support/$(name)
+	#chmod -R 777 $(DESTDIR)Library/Application\ Support/$(name)
 else
-	cp out/$(name) $(prefix)/bin/$(name)
-	mkdir -p /usr/share/$(name)
-	cp -R profile/* /usr/share/$(name)
-	chmod -R 777 /usr/share/$(name)
+	install -d $(DESTDIR)$(prefix)/bin/
+	cp out/$(name) $(DESTDIR)$(prefix)/bin/$(name)
+	install -d $(DESTDIR)usr/share/$(name)
+	cp -R profile/* $(DESTDIR)usr/share/$(name)
+	#chmod -R 777 $(DESTDIR)usr/share/$(name)
 endif
 
 uninstall:
-ifneq ($(shell id -un),root)
-	$(error "make uninstall must be run as root")
-else ifeq ($(platform),windows)
+ifeq ($(platform),windows)
 else ifeq ($(platform),macosx)
 else
-	rm $(prefix)/bin/$(name)
+	rm $(DESTDIR)$(prefix)/bin/$(name)
 endif
diff --git a/ananke/Makefile b/ananke/Makefile
index 7eb377b..b2ba0ce 100644
--- a/ananke/Makefile
+++ b/ananke/Makefile
@@ -1,7 +1,14 @@
 include ../nall/Makefile
 include ../phoenix/Makefile
 
-path := /usr/local/lib
+ifndef DESTDIR
+  DESTDIR := /
+endif
+
+ifndef prefix
+  prefix := /usr/local/lib
+endif
+
 flags := $(flags) -O3 -fomit-frame-pointer -I..
 
 all:
@@ -27,23 +34,25 @@ clean:
 install: uninstall
 ifeq ($(platform),windows)
 else ifeq ($(platform),macosx)
-	if [ ! -d ~/Library/Application\ Support/ananke ]; then mkdir ~/Library/Application\ Support/ananke; fi
-	sudo cp libananke.dylib $(path)/libananke.1.dylib
-	sudo ln -s $(path)/libananke.1.dylib $(path)/libananke.dylib
+	#if [ ! -d ~/Library/Application\ Support/ananke ]; then install -d ~/Library/Application\ Support/ananke; fi
+	install -d $(DESTDIR)$(path)
+	cp libananke.dylib $(DESTDIR)$(prefix)/libananke.1.dylib
+	ln -s $(path)/libananke.1.dylib $(DESTDIR)$(prefix)/libananke.dylib
 else
-	if [ ! -d ~/.config/ananke ]; then mkdir ~/.config/ananke; fi
-	sudo cp libananke.so $(path)/libananke.so.1
-	sudo ln -s $(path)/libananke.so.1 $(path)/libananke.so
+	#if [ ! -d ~/.config/ananke ]; then install -d ~/.config/ananke; fi
+	install -d $(DESTDIR)$(prefix)
+	cp libananke.so $(DESTDIR)$(prefix)/libananke.so.1
+	ln -s $(DESTDIR)$(path)/libananke.so.1 $(DESTDIR)$(prefix)/libananke.so
 endif
 
 uninstall:
 ifeq ($(platform),windows)
 else ifeq ($(platform),macosx)
-	if [ -f $(path)/libananke.dylib ]; then sudo rm $(path)/libananke.dylib; fi
-	if [ -f $(path)/libananke.1.dylib ]; then sudo rm $(path)/libananke.1.dylib; fi
+	if [ -f $(DESTDIR)$(path)/libananke.dylib ]; then rm $(DESTDIR)$(prefix)/libananke.dylib; fi
+	if [ -f $(DESTDIR)$(path)/libananke.1.dylib ]; then rm $(DESTDIR)$(prefix)/libananke.1.dylib; fi
 else
-	if [ -f $(path)/libananke.so ]; then sudo rm $(path)/libananke.so; fi
-	if [ -f $(path)/libananke.so.1 ]; then sudo rm $(path)/libananke.so.1; fi
+	if [ -f $(DESTDIR)$(path)/libananke.so ]; then rm $(DESTDIR)$(prefix)/libananke.so; fi
+	if [ -f $(DESTDIR)$(path)/libananke.so.1 ]; then rm $(DESTDIR)$(prefix)/libananke.so.1; fi
 endif
 
 force:
