diff --git a/higan/emulator/interface.hpp b/higan/emulator/interface.hpp
index c6cd0ce..c148f54 100755
--- a/higan/emulator/interface.hpp
+++ b/higan/emulator/interface.hpp
@@ -88,6 +88,7 @@ struct Interface {
   virtual void load(unsigned id, const stream &memory, const string &markup = "") {}
   virtual void save(unsigned id, const stream &memory) {}
   virtual void unload() {}
+  virtual void saveMemPak () {}
 
   //system interface
   virtual void connect(unsigned port, unsigned device) {}
diff --git a/higan/sfc/interface/interface.cpp b/higan/sfc/interface/interface.cpp
index 61b5198..04d40ef 100755
--- a/higan/sfc/interface/interface.cpp
+++ b/higan/sfc/interface/interface.cpp
@@ -265,6 +265,10 @@ void Interface::connect(unsigned port, unsigned device) {
   input.connect(port, (Input::Device)device);
 }
 
+void Interface::saveMemPak() {
+  system.saveMemPak();
+}
+
 void Interface::power() {
   system.power();
 }
diff --git a/higan/sfc/interface/interface.hpp b/higan/sfc/interface/interface.hpp
index 3d84500..fd3453d 100755
--- a/higan/sfc/interface/interface.hpp
+++ b/higan/sfc/interface/interface.hpp
@@ -93,6 +93,7 @@ struct Interface : Emulator::Interface {
   void load(unsigned id, const stream &stream, const string &markup = "");
   void save(unsigned id, const stream &stream);
   void unload();
+  void saveMemPak();
 
   void connect(unsigned port, unsigned device);
   void power();
diff --git a/higan/sfc/system/system.cpp b/higan/sfc/system/system.cpp
index feb48f0..e8d45df 100755
--- a/higan/sfc/system/system.cpp
+++ b/higan/sfc/system/system.cpp
@@ -165,6 +165,10 @@ void System::unload() {
   if(cartridge.has_msu1()) msu1.unload();
 }
 
+  void System::saveMemPak() {
+    if(cartridge.has_bs_slot()) bsxflash.save();
+}
+
 void System::power() {
   random.seed((unsigned)time(0));
 
diff --git a/higan/sfc/system/system.hpp b/higan/sfc/system/system.hpp
index 3bb08ae..39453d5 100755
--- a/higan/sfc/system/system.hpp
+++ b/higan/sfc/system/system.hpp
@@ -11,6 +11,7 @@ struct System : property<System> {
   void term();
   void load();
   void unload();
+  void saveMemPak();
   void power();
   void reset();
 
diff --git a/higan/target-ethos/general/presentation.cpp b/higan/target-ethos/general/presentation.cpp
index 5c5adab..baa98ec 100755
--- a/higan/target-ethos/general/presentation.cpp
+++ b/higan/target-ethos/general/presentation.cpp
@@ -163,6 +163,7 @@ void Presentation::bootstrap() {
     iEmulator->power.setText("Power");
     iEmulator->reset.setText("Reset");
     iEmulator->unload.setText("Unload");
+    iEmulator->saveMemPak.setText("Save BSX MemPak");
 
     for(auto &port : emulator->port) {
       auto iPort = new Emulator::Port;
@@ -196,10 +197,12 @@ void Presentation::bootstrap() {
     iEmulator->menu.append(iEmulator->controllerSeparator);
     if(visiblePorts == 0) iEmulator->controllerSeparator.setVisible(false);
     iEmulator->menu.append(iEmulator->unload);
+    iEmulator->menu.append(iEmulator->saveMemPak);
 
     iEmulator->power.onActivate = {&Utility::power, utility};
     iEmulator->reset.onActivate = {&Utility::reset, utility};
     iEmulator->unload.onActivate = {&Utility::unload, utility};
+    iEmulator->saveMemPak.onActivate = {&Utility::saveMemPak, utility};
 
     emulatorList.append(iEmulator);
   }
diff --git a/higan/target-ethos/general/presentation.hpp b/higan/target-ethos/general/presentation.hpp
index 1a44aa3..29e737d 100755
--- a/higan/target-ethos/general/presentation.hpp
+++ b/higan/target-ethos/general/presentation.hpp
@@ -9,6 +9,7 @@ struct Presentation : Window {
       Item power;
       Item reset;
       Item unload;
+      Item saveMemPak;
       Separator controllerSeparator;
       struct Port {
         Menu menu;
diff --git a/higan/target-ethos/utility/utility.cpp b/higan/target-ethos/utility/utility.cpp
index 1ed9441..8f089f1 100755
--- a/higan/target-ethos/utility/utility.cpp
+++ b/higan/target-ethos/utility/utility.cpp
@@ -136,6 +136,10 @@ void Utility::unload() {
   stateManager->setVisible(false);
 }
 
+void Utility::saveMemPak() {
+  system().saveMemPak();
+}
+
 void Utility::saveState(unsigned slot) {
   if(application->active == nullptr) return;
   serializer s = system().serialize();
diff --git a/higan/target-ethos/utility/utility.hpp b/higan/target-ethos/utility/utility.hpp
index dc5f3a1..639ff29 100755
--- a/higan/target-ethos/utility/utility.hpp
+++ b/higan/target-ethos/utility/utility.hpp
@@ -14,6 +14,7 @@ struct Utility {
   void reset();
   void load();
   void unload();
+  void saveMemPak();
 
   void saveState(unsigned slot);
   void loadState(unsigned slot);
